### ISSUE TITLE
scipy.integrate.dblquad gives very wrong result even for the simplest integrations

### ISSUE THREAD

**Issue by lorin-kazaz** [pyodide/pyodide#4818]
## 🐛 Bug

Using the pyodide kernel, the `scipy.integrate.dblquad` gives very wrong results even for the simplest integrations.
The integrations are correct outside of pyodide.

### To Reproduce
Enter the following code into [jupyterlite](https://jupyterlite.github.io/demo/lab/index.html)
```python
import scipy
print(scipy.__version__)
print(pyodide.__version__)

unit_square_area = scipy.integrate.dblquad(lambda y,x: 1, 0, 1, lambda x: 0, lambda x: 1)
assert abs(unit_square_area[0]-1) < unit_square_area[1], 'Unit square area calculated using scipy.integrate.dblquad of {} (+- {}) is too far from 1.0'.format(unit_square_area[0], unit_square_area[0])
```
### Expected behavior

accurate result. pass above test.

### Tested Environments

- `pyodide.__version__`: '0.25.0' with `scipy.__version__`: '1.11.2' and jupyterlite: '0.3.0' 
- `pyodide.__version__`: '0.23.0' with `scipy.__version__`: '1.9.3' and jupyterlite: '0.1.0' 
- Independent of browser. Happens in Edge (Version 124.0.2478.105 (Official build) (64-bit) and Firefox (115.11.0esr (64-bit)) and Chrome.
- Any other relevant information:

### Additional context

Related issues (unclear if related, sound similar):
- https://github.com/scipy/scipy/issues/17413 
- https://github.com/scipy/scipy/issues/6882
- https://github.com/scipy/scipy/pull/15997
- https://github.com/lfortran/lfortran/issues/1377

**Comment by hoodmane** [comment]
@lesteve maintains a list of the failing scipy tests here:
https://github.com/lesteve/scipy-tests-pyodide/blob/main/conftest.py#L40

Some day we'll get [a working fortran compiler](https://lfortran.org/) and these problems will go away. Until then, nobody has volunteered to give up their life to trying to understand what is going on with these failing tests.

**Comment by georg-rempfer** [comment]
Some more testing revealed that the issue does not only appear using dblquad but also when using nested quad integrations to manually piece together the double integration:
```
from scipy.integrate import quad

def inner_const(y):
    val = 1.0
    print(y, val)
    return val

def inner_quad(y):
    val = quad(lambda x: 1, 0, 1)[0]
    print(y, val)
    return val
```
running `quad(inner_const, 0, 1)`, which only calls quad for the outer integration produces the expected output indicating a reasonable number of function evaluations and the correct result 1.0 with low error estimate:
```0.5 1.0
0.013046735741414128 1.0
0.9869532642585859 1.0
0.06746831665550773 1.0
0.9325316833444923 1.0
0.16029521585048778 1.0
0.8397047841495122 1.0
0.2833023029353764 1.0
0.7166976970646236 1.0
0.4255628305091844 1.0
0.5744371694908156 1.0
0.002171418487095955 1.0
0.997828581512904 1.0
0.03492125432214588 1.0
0.9650787456778541 1.0
0.10959113670679155 1.0
0.8904088632932085 1.0
0.21862143266569767 1.0
0.7813785673343023 1.0
0.35280356864926987 1.0
0.6471964313507301 1.0
(1.0, 1.1102230246251565e-14)
```
running `quad(inner_quad, 0, 1)`, however, demonstrates the issue. Despite the fact that the outer quad integration gets exactly the same values from the inner function (namely 1.0 independently of y), it keeps evaluating more and more function values until eventually it logs an error and reports a completely wrong result:
```
0.5 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.25 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.75 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.125 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.375 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.625 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.875 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.0625 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.1875 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.3125 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.4375 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.5625 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.6875 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.8125 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.9375 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.03125 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.09375 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.15625 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.21875 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.28125 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.34375 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.40625 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.46875 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.53125 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.59375 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.65625 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.71875 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.78125 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.84375 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.90625 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.96875 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.015625 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.046875 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.078125 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.109375 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.140625 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.171875 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.203125 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.234375 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.265625 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.296875 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.328125 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.359375 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.390625 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.421875 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.453125 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.484375 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.515625 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.546875 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.578125 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.609375 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.640625 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.671875 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.703125 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.734375 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.765625 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
0.796875 1.0
0.013046735741414128 1.0
0.6471964313507301 1.0
0.002171418487095955 1.0
0.6471964313507301 1.0
<ipython-input-8-141f2d7cc60b>:1: IntegrationWarning: The integral is probably divergent, or slowly convergent.
  quad(inner_quad, 0, 1)
(-1.7763568394002363e-15, 1.6776703483224447e-15)
```
The only difference for the outer quad call is that the function it evaluates itself calls quad. The fact that this makes a difference despite all the returned values of the inner_* functions being identical indicates that quad in pyodide scipy is not thread-safe.

I am not familiar enough with this environment to hypothesize how this could happen, knowing that the same code works fine in CPython.

**Comment by georg-rempfer** [comment]
Evaluating the outer integral with `quad` and the inner integral with `quadrature` works properly:
```
def inner_quadrature(y):
    val = quadrature(lambda x: 1, 0, 1)[0]
    print(y, val)
    return val
```
```
0.5 1.0
0.013046735741414128 1.0
0.9869532642585859 1.0
0.06746831665550773 1.0
0.9325316833444923 1.0
0.16029521585048778 1.0
0.8397047841495122 1.0
0.2833023029353764 1.0
0.7166976970646236 1.0
0.4255628305091844 1.0
0.5744371694908156 1.0
0.002171418487095955 1.0
0.997828581512904 1.0
0.03492125432214588 1.0
0.9650787456778541 1.0
0.10959113670679155 1.0
0.8904088632932085 1.0
0.21862143266569767 1.0
0.7813785673343023 1.0
0.35280356864926987 1.0
0.6471964313507301 1.0
(1.0, 1.1102230246251565e-14)
```
I suppose `quad` and `quadrature` use separate contexts and do not interfere with each other when run in parallel.

**Comment by georg-rempfer** [comment]
I was looking for information about how scipy for pyodide is built to get a better idea about what @hoodmane means with his comment on Fortran compilers. This explains how that works right now:
https://github.com/hoodmane/pyodide/blob/main/packages/scipy/info.md

I don't think I'm ready to give up my life to dig through that and I now understand why no one else is. Fell free to close this as won't fix.

And thank you for pyodide. Using it as part of jupyter-lite is an awesome environment that allows us to collaborate with others in the following way:
1. we provide notebooks with templates for specific calculations
2. we provide environment to execute calculations
3. user provides compute power to execute them
4. user doesn't need any know-how
5. no one relies on a third party

The only other environment I could imagine serving this use-case is javascript, but I much prefer doing numerics in Python.

**Comment by lesteve** [comment]
I should take some time to contribute some kind of high-level summary of the test failures to the Pyodide repo (and maybe one day the Scipy tests I am running in https://github.com/lesteve/scipy-tests-pyodide but that's a bit more work).

I asked a question about scipy.integrate failures a while ago (one year ago already :sweat_smile: ...) in the Scientific Python Discord: https://discord.com/channels/786703927705862175/1020054666433802351/1098641451073667193 about the scipy.integrate `TestNdPPoly.test_integrate_2d` failure.

There was an answer by @ev-br saying:

> NdPPoly is cython, but the test itself compares to nquad. nquad does recursive quadpack calls, does pyodide have difficulties with fortran recursive subroutines?

I am guessing this is similar to what https://github.com/pyodide/pyodide/issues/4818#issuecomment-2138478366 mentions, `quadrature` inside `quad` works but `quad` inside `quad` does not.

IIRC "modern" Fortran construct are removed that includes `recursive` keywords because `f2c` does not know about it. This is probably the source of the bug.

Having said that, the long-term option still is a working Fortran compiler :wink:.

**Comment by ev-br** [comment]
A drive-by comment, in addition to what was already said: from a user side, you may want to use `quad_vec` or `_tanh-sinh` . IIUC has a `dbl*` version so you'd need to nest integrations manually. 

The plan to rewrite QUADPACK still exists, https://github.com/scipy/scipy/issues/18566. There was no activity on QUADPACK specifically so far, and help welcome and appreciated.

**Comment by georg-rempfer** [comment]
Can I see that Cython and F2C converted scipy code somewhere?

**Comment by hoodmane** [comment]
Looking at generated `qagse.c`, I think the problem seems pretty clear:
```C
/* Subroutine */ int dqagse_(D_fp f, doublereal *a, doublereal *b, doublereal 
	*epsabs, doublereal *epsrel, integer *limit, doublereal *result, 
	doublereal *abserr, integer *neval, integer *ier, doublereal *alist__,
	 doublereal *blist, doublereal *rlist, doublereal *elist, integer *
	iord, integer *last)
{
    /* System generated locals */
    integer i__1, i__2;
    doublereal d__1, d__2;

    /* Local variables */
    static integer k;
    static doublereal a1, a2, b1, b2;
    static integer id;
    static doublereal area;
    extern /* Subroutine */ int dqk21_(D_fp, doublereal *, doublereal *, 
	    doublereal *, doublereal *, doublereal *, doublereal *);
    static doublereal dres;
    static integer ksgn, nres;
    static doublereal area1, area2, area12;
    extern /* Subroutine */ int dqelg_(integer *, doublereal *, doublereal *, 
	    doublereal *, doublereal *, integer *);
    static doublereal small, erro12;
    static integer ierro;
    // ...
```

**Comment by hoodmane** [comment]
Indeed:
![image](https://github.com/pyodide/pyodide/assets/8739626/08c1a3a7-1957-41bb-9266-9ad9079c0674)

**Comment by hoodmane** [comment]
Thanks @georg-rempfer @lorin-kazaz @ev-br! I definitely couldn't have found this fix without your detailed investigation.

One of my main difficulties maintaining these very complicated magicks against scipy is that I have never actually used scipy or anything downstream of scipy and I have never written any fortran code. So it's very hard for me to even figure out what these functions are supposed to be doing, let alone why they might be doing it incorrectly. Perhaps with support from people domain knowledge we can make gradual progress on fixing some of the bugs.

**Comment by ev-br** [comment]
These functions are hard to figure out even with fortran knowledge and a general understanding of this corner of the maths. They come (am speculating here) straight from an era when structured programming was an abstract fad from those fancy bros from comp sci departments, far removed from the needs of real work that needs doing :-). And an "implicit save" ( == making local variables static, at a compiler's discretion) was very much not new even back then :-).

Feel free to ping me on these scipy needs btw. I can't tell heads from tails in the JS world, but (some) scipy usages I may have some idea.

**Comment by georg-rempfer** [comment]
> Looking at generated `qagse.c`, I think the problem seems pretty clear:
> 
> ```c
> /* Subroutine */ int dqagse_(D_fp f, doublereal *a, doublereal *b, doublereal 
> 	*epsabs, doublereal *epsrel, integer *limit, doublereal *result, 
> 	doublereal *abserr, integer *neval, integer *ier, doublereal *alist__,
> 	 doublereal *blist, doublereal *rlist, doublereal *elist, integer *
> 	iord, integer *last)
> {
>     /* System generated locals */
>     integer i__1, i__2;
>     doublereal d__1, d__2;
> 
>     /* Local variables */
>     static integer k;
>     static doublereal a1, a2, b1, b2;
>     static integer id;
>     static doublereal area;
>     extern /* Subroutine */ int dqk21_(D_fp, doublereal *, doublereal *, 
> 	    doublereal *, doublereal *, doublereal *, doublereal *);
>     static doublereal dres;
>     static integer ksgn, nres;
>     static doublereal area1, area2, area12;
>     extern /* Subroutine */ int dqelg_(integer *, doublereal *, doublereal *, 
> 	    doublereal *, doublereal *, integer *);
>     static doublereal small, erro12;
>     static integer ierro;
>     // ...
> ```
This looks exactly like what I imagined after I learned that old Fortran does not use a stack and that you compile to C. Awesome! I think before the "recursive" statements were added to the Fortran code, double integration would not have worked in standard scipy either. As far as I understand the Fortran code would have detected the re-entrance of the function and terminated with an error.

I think your fix @hoodmane is how one would implement the "recursive" keyword in the F2C converter. If I am right that re-entry of a function not marked as recursive in Fortran simply terminates the program with an error, then applying that fix to all functions should be ok. I certainly do not know enough about Fortran to say that with confidence, though.

Do you know how long approximately it will take until this change makes it into the main branch of https://github.com/jupyterlite/demo?

If my understanding of the history does not betray me, many of these implementations were created simultaneously with Fortran and the computers that ran them at a time when the idea of maintaining a stack that would allow functions to be used recursively by default was not mainstream, yet. I see how these few-letter Fortran functions must look like hieroglyphs unless you have gone through some scientific computing lecture with an old timer. They kept the documentation separate from their code because space on their punch cards was precious and published it as books like "Numerical Recipes" or "Handbook of Integration". I took such a lecture with a theoretical astrophysicist who made us implement all the basic routines from scratch in plain C, no includes other than stdio. That helped. Anyway, thanks again for the good work. I appreciate it.

**Comment by georg-rempfer** [comment]
I suppose this fix will end up in pyodide 0.26.1 as indicated by the milestone https://github.com/pyodide/pyodide/milestone/24
Then we have to check https://github.com/jupyterlite/pyodide-kernel/blob/v0.3.1/README.md for a version of `jupyterlite-core` that uses a version of `jupyterlite-pyodide-kernel` that uses a version of `pyodide` 0.26.1 or greater, and then we have to check when https://github.com/jupyterlite/demo/blob/main/requirements.txt uses said `jupyterlite-core` version.

Is that right?

**Comment by lesteve** [comment]
If you are deploying your own JupyterLite you can set the Pyodide version in `jupyter-lite.json`, for example this is how we do it in scikit-learn:
https://github.com/scikit-learn/scikit-learn/blob/main/doc/jupyter-lite.json

**Comment by lorin-kazaz** [comment]
Thank you guys for the help:-)

@lesteve
Just for your information: 
Lfortran seems to be ok for compiling quadpack correctly since end of 2023, according to 
[this comment](https://github.com/lfortran/lfortran/issues/2714#issuecomment-1793675784
) and PR, where they state that
> After https://github.com/lfortran/lfortran/pull/2800 gets merged, we'll be able to pass all the tests for Quadpack without any workaround/ patches.
I've pushed the latest commits here: https://github.com/certik/scipy/commits/quadpack_03


1. How valuable would it be for us if lfortran were used to compile quadpack in pyodide?
2. How reasonable would it be for us to test/deploy lfortran (instead of f2c) in pyodide?

**Comment by hoodmane** [comment]
> How valuable would it be for us if lfortran were used to compile quadpack in pyodide?

I don't really want to do a gradual switch I'd rather just wait until it can compile everything. In particular I think it will be very hard to match the f2c abi to the lfortran one. But I'd love to have people experimenting with just one library to see how it goes.

> How reasonable would it be for us to test lfortran (instead of f2c) in pyodide?

Dunno try it and find out? Last I checked there was a minor problem with -fPIC. It should be an easy fix but the code just looked already right the last time I looked into it.

https://github.com/lfortran/lfortran/issues/550

I should at least be able to hack recursion into working with the current system in the meantime.

**Comment by ilayn** [comment]
> I don't think I'm ready to give up my life to dig through that and I now understand why no one else is. Fell free to close this as won't fix.

Ehem... is this thing on? Hi everyone, I'm one of the scipy folks translating the fortran code (https://github.com/scipy/scipy/issues/18566), one lowercase character at a time. I will eventually get to `QUADPACK` but one thing we are desperately missing is folks to test the translations. Because after some time, you get blinded by all that SCREAMCASE FORTRAN CODE. Hence if you are familiar with any particular domain that uses these `jurrasicPACK` codebase we would love to get your input.

My schedule (albeit constantly disrupted by my daily work)

- `id_dist` for `scipy.linalg.interpolative`
- `PROPACK`
- `ARPACK`
- `ODRPACK`
- `MINPACK`
- `QUADPACK`
- if I am alive up to this point, I'll reconsider my life

PROPACK and ARPACK is way too important for us, (segfaults, decade-long open issues etc.) so I'd stick to those as the immediate next ones but if you think quadpack is more urgent, we can move it up.

**Comment by georg-rempfer** [comment]
> > I don't think I'm ready to give up my life to dig through that and I now understand why no one else is. Fell free to close this as won't fix.
> 
> Ehem... is this thing on? Hi everyone, I'm one of the scipy folks translating the fortran code ([scipy/scipy#18566](https://github.com/scipy/scipy/issues/18566)), one lowercase character at a time. I will eventually get to `QUADPACK` but one thing we are desperately missing is folks to test the translations. Because after some time, you get blinded by all that SCREAMCASE FORTRAN CODE. Hence if you are familiar with any particular domain that uses these `jurrasicPACK` codebase we would love to get your input.
> 
> My schedule (albeit constantly disrupted by my daily work)
> 
> * `id_dist` for `scipy.linalg.interpolative`
> * `PROPACK`
> * `ARPACK`
> * `ODRPACK`
> * `MINPACK`
> * `QUADPACK`
> * if I am alive up to this point, I'll reconsider my life
> 
> PROPACK and ARPACK is way too important for us, (segfaults, decade-long open issues etc.) so I'd stick to those as the immediate next ones but if you think quadpack is more urgent, we can move it up.

I suppose we have solved the problem at hand. We do have an interest in a well working SciPy in Pyodide, though. I have looked up what all the *PACKs do. I think I am familiar enough with MINPACK and QUADPACK to help understand the results. With work and multiple little kids, my time is quite limited, but a bit of testing like we have done here should be possible. Do you need some contact info?

**Comment by lorin-kazaz** [comment]
@hoodmane If the results of the quadpack/f2c/pyodide chain still fail some tests, please @-mention me in issues.

So I suggest you open issues for failing tests of quadpack and mention me (and georg-rempfer) in those issues, such as in https://github.com/pyodide/pyodide/pull/4822.

**Comment by georg-rempfer** [comment]
Just stumbled over some docstrings that had the literature reference for the QUADPACK documentation:

```
References
----------

.. [1] Piessens, Robert; de Doncker-Kapenga, Elise;
       Überhuber, Christoph W.; Kahaner, David (1983).
       QUADPACK: A subroutine package for automatic integration.
       Springer-Verlag.
       ISBN 978-3-540-12553-2.
```

**Comment by lorin-kazaz** [comment]
> Just stumbled over some docstrings that had the literature reference for the QUADPACK documentation:

FYI, the docstrings (and .f sources) come from here (if I understand correctly)
https://github.com/scipy/scipy/blob/main/scipy/integrate/quadpack/doc

**Comment by hoodmane** [comment]
Thanks again for the report @lorin-kazaz. The fix for this will go into Pyodide 0.26.2

**Comment by lesteve** [comment]
Great thanks a lot @hoodmane for the fix!

**Comment by georg-rempfer** [comment]
Thank's a lot! Looking forward to using this.

**Comment by ilayn** [comment]
I am not finished as we are still hunting for a small bug somewhere but https://github.com/scipy/scipy/pull/21201 gives identical results for both `inner_const` and `inner_quad`. I don't know what I did to fix it but I guess I wrote it in a safe way. It should be ready in a few days.

**Comment by georg-rempfer** [comment]
Updated to the latest version of jupyter-lite with pyodide 0.26.2 and double integration works as expected! I'm so happy. Thank you guys.

**Comment by georg-rempfer-ai** [comment]
Did you delete the whole discussion with all the tests that made us understand that Fortran77's lack of a stack was the issue? That's a shame. I found it quite instructive and I think this thread was one of the few places on the internet describing this problem.

For future reference: Fortran77 by default does not use a stack. That means everything that is a local variable in a modern language, is actually a global variable there. This behavior was carried over into Pyodide by transpiling the old Fortran code deep in SciPy from Fortran to C and compiling from C to webassembly. The double integration caused the integration function to call itself (because the integrand itself was an integration). Because the integration functions variables were all global, this completely messed up the state and it did not even converge anymore. The solution was to add a few "recursive" statements to the Fortran code, which made the integration function reentrant safe, which carried over into the transpiled C code as simple local variables.

**Comment by ilayn** [comment]
I don't see any deleted comments, maybe your GitHub UI is folding your view @georg-rempfer-ai ? For me 4 comments are folded when I load the page. But they are still there under the fold.

**Comment by georg-rempfer-ai** [comment]
You are right. I see it on the Github app now. Didn't see it on a PC earlier. Must be hidden by the UI. Sorry.
